shell/src/main/java/alluxio/cli/fs/command/PersistCommand.java

```diff
@@ -47,71 +68,116 @@ public String getCommandName() {
     return "persist";
   }
 
+  @Override
+  public Options getOptions() {
+    return new Options().addOption(PARALLELISM_OPTION);
+  }
+
   @Override
   public void validateArgs(CommandLine cl) throws InvalidArgumentException {
     CommandUtils.checkNumOfArgsNoLessThan(this, cl, 1);
   }
 
   @Override
-  protected void runPlainPath(AlluxioURI plainPath, CommandLine cl)
-      throws AlluxioException, IOException {
-    persist(plainPath);
+  public String getUsage() {
+    return "persist [-p|--parallelism <#>] <path> [<path> ...]";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Persists files or directories currently stored only in Alluxio to the UnderFileSystem.";
   }
 
   @Override
   public int run(CommandLine cl) throws AlluxioException, IOException {
+    int parallelism = 4;
+    if (cl.hasOption(PARALLELISM_OPTION.getLongOpt())) {
+      String parellismOption = cl.getOptionValue(PARALLELISM_OPTION.getLongOpt());
+      parallelism = Integer.parseInt(parellismOption);
+    }
     String[] args = cl.getArgs();
+    List<AlluxioURI> candidateUris = new ArrayList<>();
     for (String path : args) {
-      AlluxioURI inputPath = new AlluxioURI(path);
-      runWildCardCmd(inputPath, cl);
+      candidateUris.addAll(FileSystemShellUtils.getAlluxioURIs(mFileSystem, new AlluxioURI(path)));
+    }
+    final BlockingQueue<AlluxioURI> toPersist = new LinkedBlockingQueue<>();
+    for (AlluxioURI uri : candidateUris) {
+      queuePersist(mFileSystem.getStatus(uri), toPersist);
+    }
+    System.out.println("Found " + toPersist.size() + " files to persist.");
+    ExecutorService service =
+        Executors.newFixedThreadPool(parallelism, ThreadFactoryUtils.build("persist-cli-%d", true));
+    final Object progressLock = new Object();
+    AtomicInteger completedFiles = new AtomicInteger(0);
+    List<Future<Void>> futures = new ArrayList<>(parallelism);
+    for (int i = 0; i < parallelism; i++) {
+      futures.add(service.submit(new PersistCallable(toPersist, completedFiles, progressLock)));
+    }
+    try {
+      for (Future<Void> future : futures) {
+        future.get();
+      }
+    } catch (ExecutionException e) {
+      System.out.println("Fatal error: " + e);
+      return -1;
+    } catch (InterruptedException e) {
+      System.out.println("Persist interrupted, exiting.");
+      return -1;
+    } finally {
+      service.shutdownNow();
     }
     return 0;
   }
 
+  private void queuePersist(URIStatus status, BlockingQueue<AlluxioURI> toPersist)
```

Comment:
nit: could be more explicit with something like `getUnpersistedFilesRecursive`

Reply:
Updated, thanks.